File: ios/TXIMManager.m
//
//  TXIMManager.m
//  TencentIm
//
//  Created by Choi Wing Chiu on 22/8/2020.
//  Copyright © 2020 Facebook. All rights reserved.
//

#import "TXIMManager.h"
#import "TXIMMessageInfo.h"
#import "TXIMMessageBuilder.h"
#import <React/RCTLog.h>
#import <ImSDK/ImSDK.h>

@interface TXIMManager() <V2TIMSDKListener>

@end

@implementation TXIMManager {
    BOOL isInit;
    
    int sdkAppId;
    
    NSString *businessID;
    
    V2TIMConversation *conversation;
    
    NSString *currentReceiver;
    
    NSData *deviceToken;
    
    NSDictionary *configDict;
    
    NSMutableArray<V2TIMConversation *> *convLists;
    
    id<V2TIMConversationListener> conversationListener;
    
    id<V2TIMSimpleMsgListener> simpleMessageListener;
    
    id<V2TIMAdvancedMsgListener> advancedMsgListener;
    
    id<V2TIMSignalingListener> signalingListener;
}

#pragma mark - public method

+ (instancetype)getInstance {
    __strong static TXIMManager *instance;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [self new];
    });
    return instance;
}

- (instancetype)init
{
    self = [super init];
    if (self) {
        convLists = [[NSMutableArray alloc] init];
    }
    return self;
}

- (void)configDeviceToken:(NSData *)token {
    deviceToken = token;
    [self configAppAPNSDeviceToken];
}

- (void)configBusinessID:(NSString *)token {
    businessID = token;
}

- (BOOL)initWithAppId:(NSString* )appId {
    if (isInit) {
      return YES;
    }
    
    V2TIMSDKConfig *config = [[V2TIMSDKConfig alloc] init];
    config.logLevel = V2TIM_LOG_NONE;
    
    [[V2TIMManager sharedInstance] setConversationListener:nil];
    
    return [[V2TIMManager sharedInstance] initSDK:[appId intValue] config:config listener:self];
}

- (void)setConversationListener:(id <V2TIMConversationListener>)listener {
    [[V2TIMManager sharedInstance] setConversationListener:listener];
    conversationListener = listener;
}

- (void)setSimpleMessageListener:(id <V2TIMSimpleMsgListener>)listener {
    [[V2TIMManager sharedInstance] addSimpleMsgListener:listener];
    simpleMessageListener = listener;
}

- (void)setAdvancedMsgListener:(id<V2TIMAdvancedMsgListener>)listener {
    [[V2TIMManager sharedInstance] addAdvancedMsgListener:listener];
    advancedMsgListener = listener;
}

- (void)setSignalingListener:(id<V2TIMSignalingListener>)listener{
    [[V2TIMManager sharedInstance] addSignalingListener:listener];
    signalingListener = listener;
}

- (void)loginWithIdentify:(NSString *)identify
                userSig:(NSString *)userSig
                   succ:(V2TIMSucc)succ
                     fail:(V2TIMFail)fail {
    void (^login)(void) = ^(void) {
        [[V2TIMManager sharedInstance] login:identify userSig:userSig succ:^{
            [self configAppAPNSDeviceToken];
            succ();
        } fail:^(int code, NSString *desc) {
            fail(code, desc);
        }];
    };
    
    if ([[V2TIMManager sharedInstance] getLoginStatus] == V2TIM_STATUS_LOGINED) {
        if ([[[V2TIMManager sharedInstance] getLoginUser] isEqualToString:identify]) {
            login();
        } else {
            [[V2TIMManager sharedInstance] logout:^{
                login();
            } fail:^(int code, NSString *desc) {
                fail(code, desc);
            }];
        }
    } else {
        login();
    }
}

- (void)logoutWithSucc:(V2TIMSucc)succ fail:(V2TIMFail)fail {
    if ([[V2TIMManager sharedInstance] getLoginStatus] == V2TIM_STATUS_LOGOUT) {
        succ();
    } else {
        [[V2TIMManager sharedInstance] logout:succ fail:fail];
    }
}

- (void)getConversationWithType:(NSInteger)type
                       receiver:(NSString *)receiver
                           succ:(V2TIMMessageListSucc)succ
                           fail:(V2TIMFail)fail {
    if ([[V2TIMManager sharedInstance] getLoginStatus] == V2TIM_STATUS_LOGINED) {
        currentReceiver = receiver;
        if (type == 1) {
            [[V2TIMManager sharedInstance] getC2CHistoryMessageList:receiver count:100 lastMsg:nil succ:^(NSArray<V2TIMMessage *> *msgs) {
                succ(msgs);
            } fail:^(int code, NSString *desc) {
                fail(code, desc);
            }];
        } else {
            [[V2TIMManager sharedInstance] getGroupHistoryMessageList:receiver count:100 lastMsg:nil succ:^(NSArray<V2TIMMessage *> *msgs) {
                succ(msgs);
            } fail:^(int code, NSString *desc) {
                fail(code, desc);
            }];
        }
    } else {
        fail(-1, @"Please login");
    }
}

- (void)setMessageRead:(V2TIMMessage *)message
                  succ:(V2TIMSucc)succ
                  fail:(V2TIMFail)fail {
    if ([[message userID] isEqualToString:[conversation userID]]) {
        [[V2TIMManager sharedInstance] markC2CMessageAsRead:[conversation userID] succ:^{
            succ();
        } fail:^(int code, NSString *desc) {
            fail(code, desc);
        }];
    }
}

- (void)sendMessage:(int)type
            content:(NSString *)content
            isGroup:(BOOL)isGroup
               succ:(TXIMSendMsgSucc)succ
               fail:(TIMFail)fail {
    if ([[V2TIMManager sharedInstance] getLoginStatus] != V2TIM_STATUS_LOGINED) {
        fail(-1, @"Please login");
        return;
    }
    
    TXIMMessageInfo *msg = [TXIMMessageBuilder buildMessage:type content:content option:nil];
    msg.sender = [[V2TIMManager sharedInstance] getLoginUser];
    msg.receiver = currentReceiver;
    msg.isSelf = YES;
    
    NSMutableDictionary *dict = [[NSMutableDictionary alloc] init];
    [dict setValue:currentReceiver forKey:@"target"];
    
    V2TIMOfflinePushInfo *info = [[V2TIMOfflinePushInfo alloc] init];
    info.ext = [self convertToJsonData:dict];
    
    [[V2TIMManager sharedInstance] sendMessage:msg.msg
                                     receiver:!isGroup?currentReceiver: nil
                                      groupID:isGroup?currentReceiver: nil
                                      priority:V2TIM_PRIORITY_DEFAULT
                                onlineUserOnly:false
                               offlinePushInfo:info
                                      progress:^(uint32_t progress) {
        
    } succ:^{
        succ(msg);
    } fail:^(int code, NSString *desc) {
        fail(code, desc);
    }];
}

- (void)destroyConversationWithSucc:(V2TIMSucc)succ
                               fail:(V2TIMFail)fail {
    if (conversation) {
        [[V2TIMManager sharedInstance] deleteConversation:[conversation conversationID] succ:^{
            self->conversation = nil;
            self->currentReceiver = nil;
            succ();
        } fail:^(int code, NSString *desc) {
            fail(code, desc);
        }];
    }
}

- (int)getUnReadCount {
    return 0;
}

- (void)getConversationList:(V2TIMConversationResultSucc)succ fail:(V2TIMFail)fail {
    if ([[V2TIMManager sharedInstance] getLoginStatus] == V2TIM_STATUS_LOGINED) {
        [[V2TIMManager sharedInstance] getConversationList:0 count:50 succ:^(NSArray<V2TIMConversation *> *list, uint64_t nextSeq, BOOL isFinished) {
            [self updateConversationWithList:list];
            succ(list,nextSeq,isFinished);
        } fail:^(int code, NSString *desc) {
            fail(code, desc);
        }];
    }
}

#pragma mark - private method

- (void)configAppAPNSDeviceToken {
    if (deviceToken != nil && [[V2TIMManager sharedInstance] getLoginStatus] == V2TIM_STATUS_LOGINED) {
        V2TIMAPNSConfig *confg = [[V2TIMAPNSConfig alloc] init];
        confg.businessID = [businessID intValue];
        confg.token = deviceToken;
        [[V2TIMManager sharedInstance] setAPNS:confg succ:^{
            RCTLog(@"[TXIMManager] configAppAPNSDeviceToken #success");
        } fail:^(int code, NSString *desc) {
            RCTLog(@"[TXIMManager] configAppAPNSDeviceToken #fail reason:%@", desc);
        }];
    }
}

- (void)onConnecting {
    RCTLog(@"[TXIMManager] onConnecting");
}

- (void)onConnectSuccess {
    RCTLog(@"[TXIMManager] onConnectSuccess");
}

- (void)onConnectFailed:(int)code err:(NSString*)err {
    RCTLog(@"[TXIMManager] onConnectFailed");
}

- (void)updateConversationWithList:(NSArray *)convList {
    for (int i = 0 ; i < convList.count ; ++ i) {
        V2TIMConversation *conv = convList[i];
        BOOL isExit = NO;
        for (int j = 0; j < convLists.count; ++ j) {
            V2TIMConversation *_Conv = convLists[j];
            if ([[self getConversationIDFromConversation:_Conv] isEqualToString:[self getConversationIDFromConversation:conv]]) {
                [convLists replaceObjectAtIndex:j withObject:conv];
                isExit = YES;
                break;
            }
        }
        if (!isExit) {
            [convLists addObject:conv];
        }
    }
    
    [convLists sortUsingComparator:^NSComparisonResult(V2TIMConversation *obj1, V2TIMConversation *obj2) {
        return [obj2.lastMessage.timestamp compare:obj1.lastMessage.timestamp];
    }];
}

- (void)updateConversationWithMessage:(V2TIMMessage *)message {
    for (int j = 0; j < convLists.count; ++ j) {
        V2TIMConversation *_Conv = convLists[j];
        if ([[self getConversationIDFromConversation:_Conv] isEqualToString:[self getConversationIDFromMessage:message]]) {
            break;
        }
    }
}

- (NSString*) getConversationIDFromMessage:(V2TIMMessage *)message {
    return message.groupID ? message.groupID : message.userID;
}

- (NSString*) getConversationIDFromConversation:(V2TIMConversation *)conv {
    return conv.groupID ? conv.groupID : conv.userID;
}

- (NSMutableArray*) getConversation {
    return convLists;
}

- (NSString *)convertToJsonData:(NSDictionary *)dict
{
    NSError *error;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:dict options:NSJSONWritingPrettyPrinted error:&error];
    NSString *jsonString;

    if (!jsonData) {
        NSLog(@"%@",error);
    } else {
        jsonString = [[NSString alloc]initWithData:jsonData encoding:NSUTF8StringEncoding];
    }

    NSMutableString *mutStr = [NSMutableString stringWithString:jsonString];

    NSRange range = {0,jsonString.length};

    //去掉字符串中的空格
    [mutStr replaceOccurrencesOfString:@" " withString:@"" options:NSLiteralSearch range:range];

    NSRange range2 = {0,mutStr.length};

    //去掉字符串中的换行符
    [mutStr replaceOccurrencesOfString:@"\n" withString:@"" options:NSLiteralSearch range:range2];

    return mutStr;
}
@end

File: android/src/main/java/com/chiu/TXIMManager.java
package com.chiu.tencentim;

import android.content.Context;

import com.chiu.tencentim.message.TXIMMessageBuilder;
import com.chiu.tencentim.message.TXIMMessageInfo;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.WritableMap;
import com.tencent.imsdk.v2.V2TIMAdvancedMsgListener;
import com.tencent.imsdk.v2.V2TIMCallback;
import com.tencent.imsdk.v2.V2TIMConversation;
import com.tencent.imsdk.v2.V2TIMConversationListener;
import com.tencent.imsdk.v2.V2TIMConversationResult;
import com.tencent.imsdk.v2.V2TIMManager;
import com.tencent.imsdk.v2.V2TIMMessage;
import com.tencent.imsdk.v2.V2TIMMessageManager;
import com.tencent.imsdk.v2.V2TIMOfflinePushInfo;
import com.tencent.imsdk.v2.V2TIMSDKConfig;
import com.tencent.imsdk.v2.V2TIMSDKListener;
import com.tencent.imsdk.v2.V2TIMSendCallback;
import com.tencent.imsdk.v2.V2TIMSignalingListener;
import com.tencent.imsdk.v2.V2TIMSimpleMsgListener;
import com.tencent.imsdk.v2.V2TIMUserFullInfo;
import com.tencent.imsdk.v2.V2TIMValueCallback;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class TXIMManager {

    public static TXIMManager instance;

    private static Context appContext;

    private static Boolean isInit;

    private static String currentReceiver;

    private static V2TIMConversation conversation;

    private static String businessID;

    private static String sdkAppId;

    private List<V2TIMConversation> convLists;

    public static TXIMManager getInstance(){
        if(instance == null){
            instance = new TXIMManager();
        }
        return instance;
    }

    public void initWithAppId(Context context, String sdkAppID) {
        appContext = context;
        V2TIMSDKConfig config = new V2TIMSDKConfig();
        config.setLogLevel(V2TIMSDKConfig.V2TIM_LOG_INFO);

        V2TIMManager.getInstance().initSDK(context, Integer.parseInt(sdkAppID), config, new V2TIMSDKListener() {
            @Override
            public void onConnecting() {
                super.onConnecting();
            }

            @Override
            public void onConnectSuccess() {
                super.onConnectSuccess();
            }

            @Override
            public void onConnectFailed(int code, String error) {
                super.onConnectFailed(code, error);
            }
        });
    }

    public void configBusinessID(String token) {
        businessID = token;
    }

    public void setConversationListener(V2TIMConversationListener listener) {
        V2TIMManager.getConversationManager().setConversationListener(listener);
    }

    public void setSimpleMessageListener(V2TIMSimpleMsgListener listener) {
        V2TIMManager.getInstance().addSimpleMsgListener(listener);
    }

    public void setAdvancedMsgListener(V2TIMAdvancedMsgListener listener) {
        V2TIMManager.getMessageManager().addAdvancedMsgListener(listener);
    }

    public void setSignalingListener(V2TIMSignalingListener listener) {
        V2TIMManager.getSignalingManager().addSignalingListener(listener);
    }

    public void login(final String identify, final String userSig, final V2TIMCallback callback) {
        if (V2TIMManager.getInstance().getLoginStatus() == V2TIMManager.V2TIM_STATUS_LOGINED) {
            V2TIMManager.getInstance().logout(new V2TIMCallback() {
                @Override
                public void onError(int i, String s) {
                    callback.onError(i, s);
                }

                @Override
                public void onSuccess() {
                    V2TIMManager.getInstance().login(identify, userSig, new V2TIMCallback() {
                        @Override
                        public void onError(int i, String s) {
                            callback.onError(i, s);
                        }

                        @Override
                        public void onSuccess() {
                            callback.onSuccess();
                        }
                    });
                }
            });
        } else {
            V2TIMManager.getInstance().login(identify, userSig, new V2TIMCallback() {
                @Override
                public void onError(int i, String s) {
                    callback.onError(i, s);
                }

                @Override
                public void onSuccess() {
                    callback.onSuccess();
                }
            });
        }
    }

    public void logout(V2TIMCallback callback) {
        V2TIMManager.getInstance().logout(callback);
    }

    public void getConversation(int type, String receiver, final V2TIMValueCallback<List<V2TIMMessage>> callback) {
        if (V2TIMManager.getInstance().getLoginStatus() == V2TIMManager.V2TIM_STATUS_LOGINED) {
            currentReceiver = receiver;
            if (type == 1) {
                V2TIMManager.getMessageManager().getC2CHistoryMessageList(receiver, 100, null, new V2TIMValueCallback<List<V2TIMMessage>>() {
                    @Override
                    public void onError(int i, String s) {
                        callback.onError(i, s);
                    }

                    @Override
                    public void onSuccess(List<V2TIMMessage> v2TIMMessages) {
                        callback.onSuccess(v2TIMMessages);
                    }
                });
            } else {
                V2TIMManager.getMessageManager().getGroupHistoryMessageList(receiver, 100, null, new V2TIMValueCallback<List<V2TIMMessage>>() {
                    @Override
                    public void onError(int i, String s) {
                        callback.onError(i, s);
                    }

                    @Override
                    public void onSuccess(List<V2TIMMessage> v2TIMMessages) {
                        callback.onSuccess(v2TIMMessages);
                    }
                });
            }
        } else {
            callback.onError(-1, "Please login");
        }
    }

    public void setMessageRead(V2TIMMessage message, final V2TIMCallback listener) {
        if (conversation != null && conversation.getUserID().equals(message.getUserID())) {
            V2TIMManager.getMessageManager().markC2CMessageAsRead(message.getUserID(), new V2TIMCallback() {
                @Override
                public void onError(int i, String s) {
                    listener.onError(i, s);
                }

                @Override
                public void onSuccess() {
                    listener.onSuccess();
                }
            });
        }
    }

    public void sendMessage(int type, String content, boolean isGroup, final V2TIMSendCallback<V2TIMMessage> callback) throws JSONException {
        if (V2TIMManager.getInstance().getLoginStatus() != V2TIMManager.V2TIM_STATUS_LOGINED) {
            return;
        }
        TXIMMessageInfo msg = TXIMMessageBuilder.buildMessage(type, content);
        msg.setSender(V2TIMManager.getInstance().getLoginUser());
        msg.setReceiver(currentReceiver);
        msg.setSelf(true);

        WritableMap map = Arguments.createMap();
        map.putString("target", currentReceiver);

        V2TIMOfflinePushInfo info = new V2TIMOfflinePushInfo();
        JSONObject json = new JSONObject(map.toString());

        V2TIMManager.getMessageManager().sendMessage(msg.getMsg(), !isGroup ? currentReceiver:null, isGroup ? currentReceiver: null, V2TIMMessage.V2TIM_PRIORITY_DEFAULT, false,  info, new V2TIMSendCallback<V2TIMMessage>() {
            @Override
            public void onProgress(int i) {

            }

            @Override
            public void onError(int i, String s) {
                callback.onError(i, s);
            }

            @Override
            public void onSuccess(V2TIMMessage v2TIMMessage) {
                callback.onSuccess(v2TIMMessage);
            }
        });
    }

    public void destroyConversation(final V2TIMCallback callback) {
        if (conversation != null && V2TIMManager.getInstance().getLoginStatus() == V2TIMManager.V2TIM_STATUS_LOGINED) {
            V2TIMManager.getConversationManager().deleteConversation(conversation.getConversationID(), new V2TIMCallback() {
                @Override
                public void onError(int i, String s) {
                    callback.onError(i, s);
                }

                @Override
                public void onSuccess() {
                    callback.onSuccess();
                }
            });
        }
    }

    public int getUnReadCount() {
        return 0;
    }

    public List<V2TIMConversation> getConversation() {
        return convLists;
    }

    public void getConversationList(final V2TIMValueCallback<V2TIMConversationResult> callback) {
        if (V2TIMManager.getInstance().getLoginStatus() == V2TIMManager.V2TIM_STATUS_LOGINED) {
            V2TIMManager.getConversationManager().getConversationList(0, 50, new V2TIMValueCallback<V2TIMConversationResult>() {
                @Override
                public void onError(int i, String s) {
                    callback.onError(i, s);
                }

                @Override
                public void onSuccess(V2TIMConversationResult v2TIMConversationResult) {
                    updateConversationWithList(v2TIMConversationResult.getConversationList());
                    callback.onSuccess(v2TIMConversationResult);
                }
            });
        }
    }

    public void updateConversationWithList(List<V2TIMConversation> convList) {
        for (int i = 0 ; i < convList.size() ; ++ i) {
            V2TIMConversation conv = convList.get(i);
            boolean isExit = false;
            for (int j = 0 ; j < convLists.size() ; ++ j) {
                V2TIMConversation _Conv = convLists.get(j);
                if ( getConversationIDFromConversation(_Conv).equals(getConversationIDFromConversation(conv))) {
                    convLists.set(j, conv);
                    isExit = true;
                    break;
                }
            }
            if (!isExit) {
                convLists.add(conv);
            }
        }
        Collections.sort(convLists, new Comparator<V2TIMConversation>() {
            @Override
            public int compare(V2TIMConversation obj1, V2TIMConversation obj2) {
                return (int)obj2.getLastMessage().getTimestamp() - (int)obj1.getLastMessage().getTimestamp();
            }
        });
    }

    private String getConversationIDFromConversation(V2TIMConversation conv) {
            return conv.getGroupID() != null? conv.getGroupID() : conv.getUserID();
    }
}
